imap_processing/lo/l0/data_classes/science_direct_events.py

```diff
@@ -0,0 +1,213 @@
+    case number.
+
+    - Case 0 can either be a gold or silver triple. Gold triples do
+    not send down the TOF1 value and instead recover the TOF1 value
+    on the ground using the decompressed checksum.
+    - Cases 4, 6, 10, 12, 13 may be Bronze. If it's not a bronze,
+    the Position is not transmitted, but TOF3 is. If it is bronze, the table
+    should be used as is. If it's not bronze, position was not transmitted,
+    but TOF3 was transmitted.
+    - Cases 1, 2, 3, 5, 7, 9, 13 will always have a MODE of 0, so the same
+    fields will always be transmitted.
+
+    Bit Shifting:
+    TOF0, TOF1, TOF2, TOF3, and CKSM all must be shifted by one bit to the
+    left. All other fields do not need to be bit shifted.
+
+    The raw values are computed for L1A and will be converted to
+    engineering units in L1B.
+
+    Attributes
+    ----------
+    SHCOARSE : int
+        Spacecraft time.
+    COUNT: int
+        Number of direct events.
+    DATA: str
+        Compressed TOF Direct Event time tagged data.
+    TIME: numpy.ndarray
+        time tag for the direct event
+    ENERGY: numpy.ndarray
+        energy of the direct event ENA.
+    MODE: numpy.ndarray
+        Indication of how the data is packed.
+    TOF0: numpy.ndarray
+        Time of Flight 0 value for direct event.
+    TOF1: numpy.ndarray
+        Time of Flight 1 value for direct event.
+    TOF2: numpy.ndarray
+        Time of Flight 2 value for direct event.
+    TOF3: numpy.ndarray
+        Time of Flight 3 value for direct event.
+    CKSM: numpy.ndarray
+        This is checksum defined relative to the TOFs
+        condition for golden triples. If golden triples are below
+        a certain threshold in checksum it's considered golden, otherwise,
+        it's considered a silver triple. This is important for the compression
+        for golden triples because it's used to recover TOF1 because
+        compression scheme to save space on golden triples doesn't send
+        down TOF1 so it's recovered on the ground using the checksum
+    POS: numpy.ndarray
+        Stop position for the direct event. There are 4 quadrants
+        on the at the stop position.
+
+    Methods
+    -------
+    __init__(packet, software_vesion, packet_file_name):
+        Uses the CCSDS packet, version of the software, and
+        the name of the packet file to parse and store information about
+        the Direct Event packet data.
+    """
+
+    software_version: str
+    packet_file_name: str
+    ccsds_header: CcsdsData
+    SHCOARSE: int
+    COUNT: int
+    DATA: str
+    TIME: np.ndarray
+    ENERGY: np.ndarray
+    MODE: np.ndarray
+    TOF0: np.ndarray
+    TOF1: np.ndarray
+    TOF2: np.ndarray
+    TOF3: np.ndarray
+    CKSM: np.ndarray
+    POS: np.ndarray
+
+    def __init__(self, packet, software_version: str, packet_file_name: str):
+        """Initialize Science Direct Events Data class."""
+        set_attributes(self, packet)
```

Comment:
I made this an outside function because I am going to re-use it for every data class. Do you think it would still be better within the class?

Reply:
Yes, I personally think so. One other suggestion then would be to inherit your dataclasses and set this function in the base of the inheritance.

```python
class BaseLoClass:
    def _set_attributes(self, packet):
        ...

@dataclass
class ScienceDE(BaseLoClass):
    # Should have your method here

class ScienceMessageClass(BaseLoClass):
    # Another class with the same method inherited
```

Reply:
Yeah, that what it's doing prior to this PR. LoBase is currently a parent class with that function in it and the direct events data class is inheriting from it. I removed that in this PR because during a discussion about the ENA direct event data classes, I got convinced that the LoBase class isn't a true parent class of the data classes (and having that as a parent class would complicate things if ENAs are able to create parent classes for our direct events), so I created the outside function instead. 

I'm not crazy about importing that either, so I might revert back to using the LoBase and change that in the future if we start using a DE parent class across ENAs.

Reply:
😂 Sorry I missed all of these discussions! I personally think that is a good decision, and re-evaluating it later seems reasonable. Funny that you had what I suggested earlier and I just didn't realize it :)